Increasing bingo hints compatibility

DobbyDigital opened the pull request:
Made fixes to the bingo hints which allow it to be combined with more settings (useful for RSL, or people playing bingo with non-default options).

- Remove starting gear from the items to be hinted, which fixes a crash if the generator tries to find an item not in the world (allows the bingo hints to work with starting items, equipment, and songs, as well as skip_child_zelda)
- Include trial hints (still will not pop up in default bingo settings, since those have trials set to 6)
- Verify that the total number of hints does not exceed 40 (including item hints, always hints, and possibly trial hint). If that is the case, limit trial and always hints to one copy each. (allows for always hints to exceed 7, and avoid issues with trial hints)
- Return an error if the board-specific generator returns an empty list of items. Most likely scenario there is the user copied the wrong Bingosync URL, and the board was not for OoTR.

DobbyDigital committed:
Fixed starting items

-world.item_hints is checked after bingo hints are done to remove extra instances from starting items. Early testing shows it prevents the errors, but needs more testing

DobbyDigital committed:
Add comments

DobbyDigital committed:
Error handling (non-OoTR board)

Throw an error if the board provided has no item hints associated to it. Likely from  someone using a Bingosync URL not for OoTR bingo, but could potentially happen even without in rare circumstances I think?

DobbyDigital committed:
Cap hint total

Check to make sure if the total number of hints will exceed 40, if it does we limit the always and trial hints to one copy each

DobbyDigital committed:
Fixed spacing so Zannick doesn't get mad at me monkaHide

Zannick commented on Hints.py:
```diff
@@ -750,9 +755,34 @@ def buildWorldGossipHints(spoiler, world, checkedLocations=None):
+    #Skip_child_zelda can cause the same problem, but needs to be handled separately since
+    #it doesn't update the starting gear lists
+    if world.skip_child_zelda:
+        itemname = str(world.get_location('Song from Impa').item)
```
do you mean `world.get_location('Song from Impa').item.name`?

DobbyDigital replied:
Huh, I swear I tested that and it didn't work; I used that form for the other items and it worked fine, but for some reason it didn't work for me here

Just switched it though and now it seems to work fine for .... reasons?

Zannick commented on Hints.py:
```diff
@@ -750,9 +755,34 @@ def buildWorldGossipHints(spoiler, world, checkedLocations=None):
+    #This method ensures that the right number of copies are removed, e.g.
+    #if you start with one strength and hints call for two, you still get
+    #one hint for strength
+    for item in [*world.starting_items, *world.starting_equipment, *world.starting_songs]:
```
I would tend to prefer itertools.chain for this.

DobbyDigital replied:
Did not know that was a thing, but that is exactly what I was looking for here. Thanks! :D  Made the switch

Zannick commented on Hints.py:
```diff
@@ -750,9 +755,34 @@ def buildWorldGossipHints(spoiler, world, checkedLocations=None):
+
+    #Make sure the total number of hints won't pass 40. If so, we limit the always and trial hints
+    if world.hint_dist == "bingo":
+        numTrialHints=[0,1,2,3,2,1,0]
```
nit: add spaces to format this nicely.

Zannick replied:
lololol I didn't even see 290f19e when I wrote this.

DobbyDigital replied:
I knew I was going to miss one; the original version I forgot to put any of them, and then remembered you mentioning it last PR LUL

DobbyDigital committed:
Added requested fixes

- Switched `str(world.get_location('Song from Impa').item)` to `world.get_location('Song from Impa').item.name`
- Switched to using itertools.chain for iteration
- Fixed space that I missed

DobbyDigital commented:
By the way, I think GitHub is being weird because the tags were added and the commit with the changes happened at about the same time; I made the changes you requested, but GitHub is still saying there are "Changes Requested".

No idea why since I'm pretty sure I fixed everything?

Zannick commented:
This is because my review is still in the "changes requested" state, and github is waiting on me to change it. Nothing you did wrong!

DobbyDigital commented:
Phew, was worried timing for the commit had accidentally broken something xD Glad everything's fine, and carry on with the good work! POGGERS
